relay/pkg/block/push.go

```diff
@@ -114,39 +141,102 @@ func (f *Forwarder) addHeaders(headers []*btc.Header) error {
 	return f.hostChain.AddHeaders(anchorHeader.Raw, packHeaders(headers))
 }
 
-func (f *Forwarder) updateBestHeader(
-	ctx context.Context,
-	newBestHeader *btc.Header,
-) error {
-	currentBestDigest, err := f.hostChain.GetBestKnownDigest()
+func (f *Forwarder) addHeadersWithRetarget(headers []*btc.Header) error {
+	epochStart := headers[0].Height - difficultyEpochDuration
+	epochEnd := epochStart + difficultyEpochDuration - 1
+
+	oldPeriodStartHeader, err := f.btcChain.GetHeaderByHeight(epochStart)
 	if err != nil {
-		return fmt.Errorf("could not get best known digest: [%v]", err)
+		return fmt.Errorf(
+			"could not get header by height [%v]: [%v]",
+			epochStart,
+			err,
+		)
 	}
 
-	currentBestHeader, err := f.btcChain.GetHeaderByDigest(currentBestDigest)
+	oldPeriodEndHeader, err := f.btcChain.GetHeaderByHeight(epochEnd)
 	if err != nil {
 		return fmt.Errorf(
-			"could not get current best header by digest: [%v]",
+			"could not get header by height [%v]: [%v]",
+			epochEnd,
 			err,
 		)
 	}
 
-	lastCommonAncestor, err := f.findLastCommonAncestor(
-		ctx,
-		newBestHeader,
-		currentBestHeader,
+	return f.hostChain.AddHeadersWithRetarget(
+		oldPeriodStartHeader.Raw,
+		oldPeriodEndHeader.Raw,
+		packHeaders(headers),
 	)
-	if err != nil {
-		return fmt.Errorf("could not find last common ancestor: [%v]", err)
-	}
+}
 
-	limit := newBestHeader.Height - lastCommonAncestor.Height + 1
+func (f *Forwarder) updateBestHeader(
+	ctx context.Context,
+	newBestHeader *btc.Header,
+) error {
+	totalAttempts := 30
 
-	return f.hostChain.MarkNewHeaviest(
-		lastCommonAncestor.Hash,
-		currentBestHeader.Raw,
-		newBestHeader.Raw,
-		big.NewInt(limit),
+	for attempt := 1; attempt <= totalAttempts; attempt++ {
+		logger.Infof(
+			"attempt [%v] to set header [%v] as new best",
+			attempt,
+			newBestHeader.Height,
```

Comment:
Is `newBestHeader.Height` always up to date "as a new best" in the context of this loop, given that we fetch `currentBestHeader` later in this loop?